tests/reg_tests/test_jacvecprodFWD.py

```diff
@@ -0,0 +1,214 @@
+                xDvDot=xDvDot, residualDeriv=True, funcDeriv=True, fDeriv=True, mode="FD", h=step_size[key]
+            )
+
+            # the tolerances here are loose becuase different ouputs have different optimal steps
+            np.testing.assert_allclose(resDot_FD, resDot, atol=5e-5, err_msg=f"residual wrt {key}")
+
+            for func in funcsDot:
+                if np.abs(funcsDot[func]) <= 1e-16:
+                    np.testing.assert_allclose(
+                        funcsDot_FD[func], funcsDot[func], atol=5e-5, err_msg=f"{func} wrt {key}"
+                    )
+                else:
+                    np.testing.assert_allclose(
+                        funcsDot_FD[func], funcsDot[func], rtol=1e-3, err_msg=f"{func} wrt {key}"
+                    )
+
+            np.testing.assert_allclose(fDot_FD, fDot, atol=5e-7, err_msg=f"forces wrt {key}")
+
+
+@parameterized_class(test_params)
+class TestJacVecFwdCS(reg_test_classes.CmplxRegTest):
+    """
+    Tests jacobian vector products against CS.
+
+    """
+
+    N_PROCS = 2
+
+    h = 1e-40
+
+    def setUp(self):
+        if not hasattr(self, "name"):
+            # return immediately when the setup method is being called on the based class and NOT the
+            # classes created using parametrized
+            # this will happen when training, but will hopefully be fixed down the line
+            return
+        super().setUp()
+
+        options = copy.copy(adflowDefOpts)
+        options["outputdirectory"] = os.path.join(baseDir, options["outputdirectory"])
+        options.update(self.options)
+
+        self.ap = copy.deepcopy(self.aero_prob)
+
+        # Setup aeroproblem
+        self.ap = copy.deepcopy(self.aero_prob)
+        # add the default dvs to the problem
+        for dv in defaultAeroDVs:
+            self.ap.addDV(dv)
+
+        # add the default dvs to the problem
+        for dv in defaultAeroDVs:
+            self.ap.addDV(dv)
+
+        self.CFDSolver = ADFLOW_C(options=options, debug=True)
+
+        # propagates the values from the restart file throughout the code
+        self.CFDSolver.getResidual(self.ap)
+
+    # ------------------- Derivative routine checks ----------------------------
+    def cmplx_test_wDot(self):
+        if not hasattr(self, "name"):
+            # return immediately when the setup method is being called on the based class and NOT the
+            # classes created using parametrized
+            # this will happen when training, but will hopefully be fixed down the line
+            return
+
+        # perturb each input and check that the outputs match the FD to with in reason
+        wDot = self.CFDSolver.getStatePerturbation(321)
+
+        resDot_CS, funcsDot_CS, fDot_CS = self.CFDSolver.computeJacobianVectorProductFwd(
+            wDot=wDot, residualDeriv=True, funcDeriv=True, fDeriv=True, mode="CS", h=self.h
+        )
+
+        print(resDot_CS)
```

Comment:
Do we compare this again anything? I remember there was a discussion on using json files for complex tests as well. I think I defaulted to comparing against the values out of the real tests. Maybe we can do a similar thing here.

Reply:
> However, then every time we make a change to the primal code this test would fail.
Then you would re-train at that point.

> It would be preferable to call the AD routines in complex mode and directly compare the values with CS.
Can you create a real ADflow instance and test CS against AD here? 

> In practice, is it possible to complexity the ADed routines?
Yeah its doable but I don't get why that's useful here. doing CS over AD kinda gives you second derivatives. Plus its much more pain than what you are trying to do here.

Reply:
> Then you would re-train at that point.

Yeah, you could re-train. It just bothers me that AD and CS could agree but still be flagged as incorrect because they are compared using old values. 

> Yeah it's doable but I don't get why that's useful here. doing CS over AD kinda gives you second derivatives. Plus its much more pain than what you are trying to do here.

I don't want to get second derivatives, I just want to run both the master_d/b without an imaginary perturbation and the master routine with an imaginary perturbation.
That way I can have the AD and CS values in memory at the same time and I can compare them without writing either out to a file first. You can do this in TACS for example. 

With that being said. I don't think this is important enough to implement now, so I'm just going to go the reference file route.